# Heart Disease 2022 - Data Analysis and Prediction/Forcasting

# Introduction

In this notebook, we'll begin by conducting exploratory data analysis (EDA) through PostgreSQL (SQL) to explore and obtain fundamental insights into the heart disease dataset. Following this, we'll transition to Python for a detailed analysis of the Heart Disease CDC dataset. We'll adopt a systematic approach to processing, cleaning, and predicting potential indicators of heart attacks in patients. The dataset comprises of 40 columns and 445,132 rows, encompassing data on patients and their related health information.

# Our Goal

Our goal is to predict the presence of heart disease in a patient based on the given features. 
We will be using different machine learning models to predict the presence of heart disease in a patient.

# Step by Step Approach

1. Data Preprocessing 
    - Handling Missing Values
    - Handling Imbalanced Data
    - Anomaly Detection
2. Exploratory Data Analysis
3. Data Visualization
4. Feature Engineering
    - Feature Selection
    - Feature Importance
    - Feature Scaling
    - Handling Categorical Variables
    - Splitting the data into training and testing set
5. Model Building
    - Logistic Regression
    - K-Nearest Neighbors
    - Support Vector Machine
    - Decision Tree
    - Random Forest
    - XGBoost
    - Neural Network
6. Model Evaluation
    - Confusion Matrix
    - Classification Report
    - ROC-AUC Curve
    - Cross-Validation
7. Conclusion and my thoughts on improving the model

# SQL Data Analysis

### Firstly we will start off with connecting to our local database and then we will perform some basic SQL queries to get a better understanding of the data.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

%matplotlib inline

In [2]:
import os
from dotenv import load_dotenv

# Load the .env file using the load_dotenv function from the dotenv package
load_dotenv()

# Access the environment variables using the os.environ dictionary
hostname = os.environ['PGHOST']
username = os.environ['PGUSER']
password = os.environ['PGPASSWORD']
database = os.environ['PGDATABASE']
port = os.environ['PGPORT']

# Load the SQL magic extension
%load_ext sql

# Assigning the %sql magic to a variable that we will use to connect to the database
connection_string = f'postgresql://{username}:{password}@{hostname}:{port}/{database}'

# Use the %sql magic to connect to the database
%sql $connection_string

In [3]:
%%sql
-- 1. Let's start by checking the first few rows of the heart_disease table.

SELECT *
FROM heart_disease
LIMIT 5

 * postgresql://alee:***@localhost:5432/postgres
5 rows affected.


[('Alabama', 'Female', 'Very good', Decimal('0.0'), Decimal('0.0'), 'Within past year (anytime less than 12 months ago)', 'No', Decimal('8.0'), None, 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'Never smoked', 'Not at all (right now)', 'No', 'White only, Non-Hispanic', 'Age 80 or older', None, None, None, 'No', 'No', 'Yes', 'No', 'Yes, received tetanus shot but not sure what type', 'No', 'No'),
 ('Alabama', 'Female', 'Excellent', Decimal('0.0'), Decimal('0.0'), None, 'No', Decimal('6.0'), None, 'No', 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Never smoked', 'Never used e-cigarettes in my entire life', 'No', 'White only, Non-Hispanic', 'Age 80 or older', Decimal('1.6'), Decimal('68.04'), Decimal('26.57'), 'No', 'No', 'No', 'No', 'No, did not receive any tetanus shot in the past 10 years', 'No', 'No'),
 ('Alabama', 'Female', 'Very good', Decimal('2.0'), Decimal('3.0'), 'Within past year (anytime less than 12 months ago)', 'Yes', Decimal('5.0'), None, 'No', 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Never smoked', 'Never used e-cigarettes in my entire life', 'No', 'White only, Non-Hispanic', 'Age 55 to 59', Decimal('1.57'), Decimal('63.5'), Decimal('25.61'), 'No', 'No', 'No', 'No', None, 'No', 'Yes'),
 ('Alabama', 'Female', 'Excellent', Decimal('0.0'), Decimal('0.0'), 'Within past year (anytime less than 12 months ago)', 'Yes', Decimal('7.0'), None, 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'Yes', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Current smoker - now smokes some days', 'Never used e-cigarettes in my entire life', 'Yes', 'White only, Non-Hispanic', None, Decimal('1.65'), Decimal('63.5'), Decimal('23.3'), 'No', 'No', 'Yes', 'Yes', 'No, did not receive any tetanus shot in the past 10 years', 'No', 'No'),
 ('Alabama', 'Female', 'Fair', Decimal('2.0'), Decimal('0.0'), 'Within past year (anytime less than 12 months ago)', 'Yes', Decimal('9.0'), None, 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'No', 'Never smoked', 'Never used e-cigarettes in my entire life', 'Yes', 'White only, Non-Hispanic', 'Age 40 to 44', Decimal('1.57'), Decimal('53.98'), Decimal('21.77'), 'Yes', 'No', 'No', 'Yes', 'No, did not receive any tetanus shot in the past 10 years', 'No', 'No')]

In [4]:
%%sql
-- 2. Check the number of males and females in the column (sex)

SELECT sex, COUNT(*)
FROM heart_disease
GROUP BY 1

 * postgresql://alee:***@localhost:5432/postgres
2 rows affected.


[('Female', 235893), ('Male', 209239)]

In [5]:
%%sql
-- 3. Height, weight, bmi for each sex

SELECT  sex, ROUND(MAX(heightinmeters), 2) AS max_height, ROUND(MIN(heightinmeters), 2) AS min_height, ROUND(AVG(heightinmeters),2) AS avg_height, 
        ROUND(MAX(weightinkilograms),2) AS max_weight, ROUND(MIN(weightinkilograms),2) AS min_weight, ROUND(AVG(weightinkilograms),2) AS avg_weight, 
        ROUND(MAX(weightinkilograms/(heightinmeters^2)), 2) AS max_bmi, ROUND(MIN(weightinkilograms/(heightinmeters^2)), 2) AS min_bmi,
        ROUND(AVG(weightinkilograms/(heightinmeters^2)), 2) AS avg_bmi
FROM heart_disease
GROUP BY 1
ORDER BY 1 DESC

 * postgresql://alee:***@localhost:5432/postgres
2 rows affected.


[('Male', Decimal('2.41'), Decimal('0.91'), Decimal('1.78'), Decimal('292.57'), Decimal('22.68'), Decimal('90.76'), Decimal('173.80'), Decimal('6.77'), Decimal('28.60')),
 ('Female', Decimal('2.41'), Decimal('0.91'), Decimal('1.63'), Decimal('290.30'), Decimal('22.68'), Decimal('75.83'), Decimal('235.53'), Decimal('7.16'), Decimal('28.47'))]

Strangely, we can see that the disparity between height and weight of both sex is relatively small, with the both the sexes maximum height being 2.41 meters and the maximum weight being around 290 kilograms, we get an average BMI of 28. Which is fairly high and is considered overweight.

In [6]:
%%sql

-- 4. The average age of patients with heart disease

SELECT DISTINCT(agecategory), COUNT(*)
FROM heart_disease
GROUP BY agecategory
ORDER BY agecategory

 * postgresql://alee:***@localhost:5432/postgres
14 rows affected.


[('Age 18 to 24', 26941),
 ('Age 25 to 29', 21990),
 ('Age 30 to 34', 25807),
 ('Age 35 to 39', 28526),
 ('Age 40 to 44', 29942),
 ('Age 45 to 49', 28531),
 ('Age 50 to 54', 33644),
 ('Age 55 to 59', 36821),
 ('Age 60 to 64', 44511),
 ('Age 65 to 69', 47099),
 ('Age 70 to 74', 43472),
 ('Age 75 to 79', 32518),
 ('Age 80 or older', 36251),
 (None, 9079)]

In [7]:
%%sql

-- 5. Ranking the average BMI of patients by age category and only show the highest average BMI for each age category

WITH CTE AS (
    SELECT sex, agecategory, ROUND(AVG(bmi)) as avg_bmi, RANK() OVER(ORDER BY ROUND(AVG(bmi)) DESC) AS bmi_rank
    FROM heart_disease
    WHERE bmi IS NOT NULL AND agecategory IS NOT NULL
    GROUP BY sex, agecategory
)

SELECT *
FROM CTE
WHERE bmi_rank = 1
ORDER BY sex DESC, agecategory, bmi_rank

 * postgresql://alee:***@localhost:5432/postgres
6 rows affected.


[('Male', 'Age 45 to 49', Decimal('30'), 1),
 ('Male', 'Age 50 to 54', Decimal('30'), 1),
 ('Male', 'Age 55 to 59', Decimal('30'), 1),
 ('Female', 'Age 40 to 44', Decimal('30'), 1),
 ('Female', 'Age 45 to 49', Decimal('30'), 1),
 ('Female', 'Age 50 to 54', Decimal('30'), 1)]

- As we can see here, the average bmi of both males and females is 30 and are between ages 40-60, which is difinitely considered obese. This is a major risk factor for heart disease.
- Now lets check which sex and age group has the largest number of patients with heart attacks.

In [8]:
%%sql

-- 6. Checking for the number of patients with heart attacks and the number of patients without heart attacks
--    between ages 45 to 59 (patients with high average BMI)

WITH NESTED_CTE AS (

    -- CTE to get the number of patients with heart attacks and the number of patients without heart attacks
    -- between the ages of 45 to 59
    WITH CTE AS (
        SELECT sex, agecategory, hadheartattack, count(*) as number_of_patients
        FROM heart_disease
        WHERE (agecategory = 'Age 45 to 49' OR agecategory = 'Age 50 to 54' OR agecategory = 'Age 55 to 59')
                AND hadheartattack IS NOT NULL
        GROUP BY sex, agecategory, hadheartattack
    )

    -- This query to get the lag of the number of patients with heart attacks and the number of patients without heart attacks
    -- based on sex and age category
    SELECT *, LAG(number_of_patients, 1, 0) OVER(PARTITION BY sex ORDER BY agecategory, hadheartattack) AS LAG
    FROM CTE
    GROUP BY CTE.sex, CTE.agecategory, CTE.hadheartattack, CTE.number_of_patients

)

-- Here we calculate the percentage of patients with heart attacks based on the number of patients 
-- with heart attacks versus the number of patients without heart attacks within the same age category and sex
SELECT  NESTED_CTE.sex, NESTED_CTE.agecategory, NESTED_CTE.hadheartattack, NESTED_CTE.number_of_patients,
        CASE WHEN number_of_patients < LAG THEN ROUND((number_of_patients / (CAST(LAG AS numeric) + number_of_patients)) * 100, 2) 
        ELSE 0 END AS perc_per_age
FROM NESTED_CTE
WHERE hadheartattack = 'Yes'
ORDER BY perc_per_age DESC, agecategory



 * postgresql://alee:***@localhost:5432/postgres
6 rows affected.


[('Male', 'Age 55 to 59', 'Yes', 1180, Decimal('6.86')),
 ('Male', 'Age 50 to 54', 'Yes', 748, Decimal('4.79')),
 ('Female', 'Age 55 to 59', 'Yes', 780, Decimal('4.02')),
 ('Male', 'Age 45 to 49', 'Yes', 423, Decimal('3.14')),
 ('Female', 'Age 50 to 54', 'Yes', 502, Decimal('2.81')),
 ('Female', 'Age 45 to 49', 'Yes', 305, Decimal('2.04'))]

- While grouping the data into sex and age groups, we can see that the highest number of patients with heart attacks are males between the ages of 55-59. This is a very important insight as it can help us in predicting the presence of heart disease in a patient.
- We can also interpreolate from the information above that as the age increases (whether male or female), the number of patients with heart attacks also increases. This is a very important insight as it can help us in predicting the presence of heart disease in a patient.

# Heart Disease Dataset with Nans and Outliers

Now, lets start doing some in-depth analysis of the dataset using Python. We will first start off by loading the dataset and then we will perform some basic data preprocessing (handling missing data, handling outliers, etc) and then we will move on to the exploratory data analysis.

In [9]:
# Read in the heart_disease table from the database
df = pd.read_sql_table('heart_disease', connection_string)

In [10]:
df.head(5)

state     sex generalhealth  physicalhealthdays  mentalhealthdays  \
0  Alabama  Female     Very good                 0.0               0.0   
1  Alabama  Female     Excellent                 0.0               0.0   
2  Alabama  Female     Very good                 2.0               3.0   
3  Alabama  Female     Excellent                 0.0               0.0   
4  Alabama  Female          Fair                 2.0               0.0   

                                     lastcheckuptime physicalactivities  \
0  Within past year (anytime less than 12 months ...                 No   
1                                               None                 No   
2  Within past year (anytime less than 12 months ...                Yes   
3  Within past year (anytime less than 12 months ...                Yes   
4  Within past year (anytime less than 12 months ...                Yes   

   sleephours removedteeth hadheartattack  ... heightinmeters  \
0         8.0         None             No  ...            NaN   
1         6.0         None             No  ...           1.60   
2         5.0         None             No  ...           1.57   
3         7.0         None             No  ...           1.65   
4         9.0         None             No  ...           1.57   

  weightinkilograms    bmi alcoholdrinkers hivtesting fluvaxlast12  \
0               NaN    NaN              No         No          Yes   
1             68.04  26.57              No         No           No   
2             63.50  25.61              No         No           No   
3             63.50  23.30              No         No          Yes   
4             53.98  21.77             Yes         No           No   

  pneumovaxever                                  tetanuslast10tdap  \
0            No  Yes, received tetanus shot but not sure what type   
1            No  No, did not receive any tetanus shot in the pa...   
2            No                                               None   
3           Yes  No, did not receive any tetanus shot in the pa...   
4           Yes  No, did not receive any tetanus shot in the pa...   

  highrisklastyear covidpos  
0               No       No  
1               No       No  
2               No      Yes  
3               No       No  
4               No       No  

[5 rows x 40 columns]

In [11]:
# Get a brief summary of the dataframe, column data types and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 445132 entries, 0 to 445131
Data columns (total 40 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   state                      445132 non-null  object 
 1   sex                        445132 non-null  object 
 2   generalhealth              443934 non-null  object 
 3   physicalhealthdays         434205 non-null  float64
 4   mentalhealthdays           436065 non-null  float64
 5   lastcheckuptime            436824 non-null  object 
 6   physicalactivities         444039 non-null  object 
 7   sleephours                 439679 non-null  float64
 8   removedteeth               433772 non-null  object 
 9   hadheartattack             442067 non-null  object 
 10  hadangina                  440727 non-null  object 
 11  hadstroke                  443575 non-null  object 
 12  hadasthma                  443359 non-null  object 
 13  hadskincancer              44

## Train Test Split and missing Target Variable removal before preprocessing to avoid data leakage

In [12]:
# Check if our target variable has any missing data and remove the rows with missing data
print(f"Missing Data in Target: {df['hadheartattack'].isnull().sum()}")

# Drop rows with missing data
df.dropna(subset=['hadheartattack'], inplace=True)

# Check for missing data in the dataframe
print(f"Missing Data in Target after removal: {df['hadheartattack'].isnull().sum()}")

Missing Data in Target: 3065
Missing Data in Target after removal: 0


In [13]:
from sklearn.model_selection import train_test_split

# Here we split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df.drop('hadheartattack', axis=1), df['hadheartattack'], test_size=0.2, random_state=42)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

# Split the training set into training and validation sets
X_train_new, X_val, y_train_new, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42)
print(X_train_new.shape, X_val.shape, y_train_new.shape, y_val.shape)

(353653, 39) (88414, 39) (353653,) (88414,)
(282922, 39) (70731, 39) (282922,) (70731,)


## Preprocessing Step - Imputation of Missing Values and Handling Outliers

### Checking for Duplications

In [14]:
# Using the duplicated method to find the sum of duplicate rows
print(f'Number of Duplicate Rows: {X_train.duplicated().sum()}')
# Here we find the percentage of duplicate rows versus the entire dataset
print(f'Percentage of Duplicate Rows versus entire dataset: {X_train.duplicated().sum()/len(X_train)*100:.2f}%')
X_train[X_train.duplicated(keep=False)]

Number of Duplicate Rows: 86
Percentage of Duplicate Rows versus entire dataset: 0.02%


state     sex generalhealth  physicalhealthdays  \
48525   Connecticut    Male     Excellent                 0.0   
66741       Florida    Male     Very good                 0.0   
353458        Texas    Male     Very good                 0.0   
265117     New York    Male          Good                 0.0   
411070   Washington    Male     Very good                 0.0   
...             ...     ...           ...                 ...   
208066    Minnesota  Female     Very good                 0.0   
358393        Texas    Male     Very good                 0.0   
52385   Connecticut  Female     Very good                 0.0   
34866    California    Male     Excellent                 0.0   
275320     New York    Male          Good                 0.0   

        mentalhealthdays                                    lastcheckuptime  \
48525                0.0  Within past year (anytime less than 12 months ...   
66741                0.0  Within past year (anytime less than 12 months ...   
353458               0.0  Within past year (anytime less than 12 months ...   
265117               0.0  Within past year (anytime less than 12 months ...   
411070               0.0  Within past year (anytime less than 12 months ...   
...                  ...                                                ...   
208066               0.0  Within past year (anytime less than 12 months ...   
358393               0.0  Within past year (anytime less than 12 months ...   
52385                0.0  Within past year (anytime less than 12 months ...   
34866                0.0  Within past year (anytime less than 12 months ...   
275320               0.0  Within past year (anytime less than 12 months ...   

       physicalactivities  sleephours  removedteeth hadangina  ...  \
48525                 Yes         6.0  None of them        No  ...   
66741                 Yes         8.0  None of them        No  ...   
353458                Yes         8.0  None of them        No  ...   
265117                Yes         8.0  None of them        No  ...   
411070                Yes         7.0  None of them        No  ...   
...                   ...         ...           ...       ...  ...   
208066                 No         8.0  None of them        No  ...   
358393                Yes         7.0  None of them        No  ...   
52385                 Yes         7.0  None of them        No  ...   
34866                 Yes         7.0  None of them        No  ...   
275320                Yes         8.0  None of them        No  ...   

       heightinmeters weightinkilograms   bmi alcoholdrinkers hivtesting  \
48525             NaN               NaN   NaN            None       None   
66741             NaN               NaN   NaN            None       None   
353458            NaN               NaN   NaN            None       None   
265117            NaN               NaN   NaN            None       None   
411070           1.93            104.33  28.0             Yes        Yes   
...               ...               ...   ...             ...        ...   
208066            NaN               NaN   NaN            None       None   
358393            NaN               NaN   NaN            None       None   
52385             NaN               NaN   NaN            None       None   
34866             NaN               NaN   NaN            None       None   
275320            NaN               NaN   NaN            None       None   

       fluvaxlast12 pneumovaxever  \
48525          None          None   
66741          None          None   
353458         None          None   
265117         None          None   
411070          Yes           Yes   
...             ...           ...   
208066         None          None   
358393         None          None   
52385          None          None   
34866          None          None   
275320         None          None   

                                        tetanuslast10tdap highrisklastyear  \
48525  

As we can see, there is a low number of duplicates within the dataset. There are 102 rows of duplicates out of 356105. Yet, removing them will unlikely have a significant impact on the overall dataset. The duplicates could most likely be real data, where two individuals out of 356105 are similar, since there are no unique characteristics to identify an individual, and removing them could result in a minute loss of valuable information. Therefore, we will not remove the duplicates.

### Getting a feeling on how much data is missing

In [15]:
# Here we can see the percentage of missing values in each column
perc_missing_vals = (X_train.isnull().sum() / len(X_train) * 100).round(3)

# Here we find the largest and smallest percentage of missing values
print(f'Largest Percentage of Missing: {perc_missing_vals.max()}%')
print(f'Smallest Percentage of Missing: {perc_missing_vals[perc_missing_vals > 0].min()}%')
print(f'There are {perc_missing_vals[perc_missing_vals > 0].count()} out of {len(X_train.columns)} columns that have missing values')

Largest Percentage of Missing: 18.397%
Smallest Percentage of Missing: 0.179%
There are 37 out of 39 columns that have missing values


## Seperate the dataset into numerical and categorical features

In [16]:
# Here we created a function to easily separate the columns into numeric and categorical
# Select_dtypes is a method in pandas that is used to select columns based on their data type
# We can use the parameter include to specify the data types we want to include in the result
def sep_column_types(dataframe):
    numeric_columns = dataframe.select_dtypes(include=[np.number])
    categorical_columns = dataframe.select_dtypes(include=[object])
    return numeric_columns, categorical_columns

# Initialized the function and assigned the results to the variables: numeric_columns and categorical_columns
numeric_columns, categorical_columns = sep_column_types(X_train)

In [17]:
# Print out the columns and the number of columns in each dataframe based on the data type

# Numeric Columns
print(numeric_columns.columns)
# Length of Numeric Columns
print(len(numeric_columns.columns))
# Categorical Columns
print(categorical_columns.columns)
# Length of Categorical Columns
print(len(categorical_columns.columns))
# Total Columns
print(X_train.columns)
# Length of Total Columns
print(len(X_train.columns))

Index(['physicalhealthdays', 'mentalhealthdays', 'sleephours',
       'heightinmeters', 'weightinkilograms', 'bmi'],
      dtype='object')
6
Index(['state', 'sex', 'generalhealth', 'lastcheckuptime',
       'physicalactivities', 'removedteeth', 'hadangina', 'hadstroke',
       'hadasthma', 'hadskincancer', 'hadcopd', 'haddepressivedisorder',
       'hadkidneydisease', 'hadarthritis', 'haddiabetes',
       'deaforhardofhearing', 'blindorvisiondifficulty',
       'difficultyconcentrating', 'difficultywalking',
       'difficultydressingbathing', 'difficultyerrands', 'smokerstatus',
       'ecigaretteusage', 'chestscan', 'raceethnicitycategory', 'agecategory',
       'alcoholdrinkers', 'hivtesting', 'fluvaxlast12', 'pneumovaxever',
       'tetanuslast10tdap', 'highrisklastyear', 'covidpos'],
      dtype='object')
33
Index(['state', 'sex', 'generalhealth', 'physicalhealthdays',
       'mentalhealthdays', 'lastcheckuptime', 'physicalactivities',
       'sleephours', 'removedteeth', 'hadangi

## Checking the Target Variable for imbalance: Heart Disease (0 = no, 1 = yes)

In [18]:
y_train.isnull().sum()

0

In [19]:
y_train.value_counts(normalize=True)

hadheartattack
No     0.943501
Yes    0.056499
Name: proportion, dtype: float64

- As we can see, the target variable is imbalanced. We will have to handle this imbalance before training our data on certain models.
- We will use SMOTE to handle the imbalance in the target variable. SMOTE is an oversampling technique that generates synthetic samples from the minority class.

## <font color="yellow">Categorical Columns</font>

### Imputing numerical nulls and removing the outliers

In [20]:
cat_list = list(categorical_columns.columns)

In [21]:
num_list = list(numeric_columns.columns)

In [22]:
from sklearn.impute import SimpleImputer
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

# Here we create a pipeline to impute missing values in the numeric columns and categorical columns
ct = make_column_transformer(
    (SimpleImputer(strategy='median'), X_train.select_dtypes(include=[np.number]).columns),
    (SimpleImputer(missing_values=pd.NA, strategy='most_frequent'), X_train.select_dtypes(include=[object]).columns),
    remainder='passthrough'
)

# Here we will create the pipeline for the transformation
pipe = make_pipeline(ct)

# Fit the pipeline to the training data
X_train_transformed = pipe.fit_transform(X_train)

# Here we convert the transformed data into a dataframe
# Although we were unable to retain the column dtypes due to the transformation and pandas dataframe infering the dtypes
X_train_transformed = pd.DataFrame(X_train_transformed, columns=num_list + cat_list)

# Here we check for missing values in the transformed data
# X_train_transformed.isnull().sum()


## <font color="yellow">Numerical Columns</font>

For the numerical columns, we want to manually impute the nulls and remove the outliers. We will use the a variety of statistical methods to impute the nulls based on the distribution of the column data.

In [ ]:
from sklearn.impute import SimpleImputer

### 1. Physial Health Days

In [ ]:
# Create a function to get a quick summary of the missing values in the dataframe

def missing_summary(dataframe, column_name):
    # Checking for the number of nulls in the column
    print(f"Number of Nulls in {column_name}: {dataframe[column_name].isnull().sum()}")
    # Percentage of nulls in column
    dataframe[column_name].isnull().sum()/len(dataframe[column_name])
    # Brief statistical summary of the column
    print(f"Describe: \n{dataframe[column_name].describe()}")

# Call the function and pass the physicalhealthdays column to it
missing_summary(numeric_columns, 'physicalhealthdays')

In [ ]:
# Create a function to visualize the distribution of the column

def plot_distribution(dataframe, column_name):
    fig, ax = plt.subplots(1, 2, figsize=(15, 5))
    sns.histplot(dataframe[column_name], kde=True, ax=ax[0])
    sns.boxplot(x=dataframe[column_name], ax=ax[1])
    ax[0].set_title(f'{column_name} Histogram')
    ax[1].set_title(f'{column_name} Boxplot')
    plt.show()

# Call the function and pass the physicalhealthdays column to it
plot_distribution(numeric_columns, 'physicalhealthdays')

### Impuation of nulls and removal of outliers for the physical health days column

In [ ]:
# Fill the nulls with the mode because the distribution is skewed and mode is not affected by outliers (discrete values)
# Used index 0 because the mode function returns a series rather than a single value
imputer = SimpleImputer(strategy='most_frequent')

# Fit the imputer to the column
imputer.fit(numeric_columns['physicalhealthdays'].values.reshape(-1, 1))

# Transform the column
numeric_columns['physicalhealthdays'] = imputer.transform(numeric_columns['physicalhealthdays'].values.reshape(-1, 1))

# Check if the nulls have been filled
print(f"Number of Nulls in physicalhealthdays: {numeric_columns['physicalhealthdays'].isnull().sum()}")

### 2. Mental Health Days

In [ ]:
missing_summary(numeric_columns, 'mentalhealthdays')

In [ ]:
plot_distribution(numeric_columns, 'mentalhealthdays')

In [ ]:
# Create an instance of SimpleImputer
imputer = SimpleImputer(strategy='most_frequent')

# Fit the imputer to the data
imputer.fit(numeric_columns[['mentalhealthdays']])

# Transform the data by replacing missing values with the mean
numeric_columns['mentalhealthdays'] = imputer.transform(numeric_columns[['mentalhealthdays']])

# Check if the nulls have been filled
print(f"Number of Nulls in physicalhealthdays: {numeric_columns['mentalhealthdays'].isnull().sum()}")

### 3. Sleep Hours

In [ ]:
numeric_columns['SleepHours'].isnull().sum()

In [ ]:
numeric_columns['SleepHours'].describe()

In [ ]:
from scipy.stats import boxcox

# Apply Box-Cox transformation to the SleepHours column
sns.displot(boxcox(numeric_columns['SleepHours'], lmbda=0), kde=True)



In [ ]:
numeric_columns['SleepHours'].plot(kind='kde')

In [ ]:
numeric_columns['SleepHours'].plot(kind='hist', bins=20)

In [ ]:
numeric_columns['SleepHours'].describe()

In [ ]:
numeric_columns['SleepHours'].plot(kind='box')

In [ ]:
numeric_columns['SleepHours'].median()

In [ ]:
# We could either impute the mean or median for the null values in the SleepHours column because 
# the 50% quratile is 7(median) and the mean is 7.022 (w/o the nulls)

# Impute the nulls with the median
numeric_columns['SleepHours'] = SimpleImputer(strategy='median').fit_transform(numeric_columns[['SleepHours']])
numeric_columns['SleepHours'].isnull().sum()

In [ ]:
from scipy.stats import zscore
from scipy.stats.mstats import winsorize

# Here we utilize the z-score to identify the outliers and then winsorize the data 
# (bounding the outliers to the extremes of the distribution to reduce their impact)
# In addition, we have a normal distribution so we can use the z-score to identify the outliers

def remove_outlier_zscore(df, column):
    abs_z_scores = np.abs(zscore(df[column]))
    filtered_entries = (abs_z_scores < 3)
    return filtered_entries

def non_remove_outlier_winsor(df, column):
    winsorized_column = winsorize(df[column], limits=[0.05, 0.05])
    return winsorized_column

zscore_remove_outlier = remove_outlier_zscore(numeric_columns, 'SleepHours')
winsorize_data = non_remove_outlier_winsor(numeric_columns, 'SleepHours')

fig, ax = plt.subplots(1, 2, figsize=(15, 5))
sns.boxplot(numeric_columns['SleepHours'][zscore_remove_outlier], ax=ax[0])
sns.boxplot(winsorize_data, ax=ax[1])
ax[0].set_title('Z-Score Remove Outlier')
ax[1].set_title('Winsorize Data')
ax[0].text(0.5, 0.95, f'Shape: {numeric_columns["SleepHours"][zscore_remove_outlier].shape}', ha='center', va='center', transform=ax[0].transAxes, bbox=dict(facecolor='white', edgecolor='black'))
ax[1].text(0.5, 0.95, f'Shape: {winsorize_data.shape}', ha='center', va='center', transform=ax[1].transAxes, bbox=dict(facecolor='white', edgecolor='black'))


# numeric_columns['SleepHours'][zscore_remove_outlier]
# numeric_columns['SleepHours'][winsorize_data]


### 4. Height in meters

In [ ]:
numeric_columns['HeightInMeters'] = SimpleImputer(strategy='mean').fit_transform(numeric_columns[['HeightInMeters']])

In [ ]:
numeric_columns['HeightInMeters'].plot(kind='hist')

### 5. Weight in kilo

In [ ]:
numeric_columns['WeightInKilograms'].plot(kind='hist')

In [ ]:
numeric_columns['WeightInKilograms'] = SimpleImputer(strategy='median').fit_transform(numeric_columns[['WeightInKilograms']])

In [ ]:
numeric_columns['WeightInKilograms'].plot(kind='hist')

### Boxcox used to normalize the distribution of the numerical features - lambda = 0 (logarithm)

In [ ]:
numeric_columns['WeightInKilograms'] = boxcox(numeric_columns['WeightInKilograms'], lmbda=0)

In [ ]:
numeric_columns['WeightInKilograms'].plot(kind='hist')

### 6. BMI

In [ ]:
numeric_columns['BMI'].plot(kind='hist')

In [ ]:
numeric_columns['BMI'] = SimpleImputer(strategy='median').fit_transform(numeric_columns[['BMI']])

In [ ]:
numeric_columns['BMI'].plot(kind='hist')

### Boxcox used to normalize the distribution of the numerical features - lambda = 0 (logarithm)

In [ ]:
numeric_columns['BMI'] = boxcox(numeric_columns['BMI'], lmbda=0)

In [ ]:
numeric_columns['BMI'].plot(kind='hist')

### Numerical Features (Checking for Nulls and Outliers)

In [ ]:
numeric_columns.isnull().sum()

## Read from Postgres Database

In [ ]:
import pandas as pd
from sqlalchemy import create_engine
from sqlalchemy import inspect
from sqlalchemy.orm import Session
engine = create_engine(connection_string)
conn = engine.connect()
inspector = inspect(engine)
print(inspector.get_table_names())

In [ ]:
df = pd.read_sql("SELECT * FROM heart_disease", conn)
df

In [ ]:
# Checking for nulls - No nulls
df.isnull().sum()

In [ ]:
# Columns: 40, Rows: 246022
df.info()

In [ ]:
df['hadheartattack'].value_counts(normalize=True)

In [ ]:
# Create list of columns
columns_list = list(df.columns)
columns_list

In [ ]:
object_columns = df.select_dtypes(include='object')
numeric_columns = df.select_dtypes(exclude='object')

In [ ]:
object_columns.columns

In [ ]:
numeric_columns.columns

In [ ]:
df['generalhealth'].value_counts().plot(kind='barh', title='General Health')

## Train Test Split

In [ ]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import LabelEncoder

In [ ]:
encoder = LabelEncoder()

for column in object_columns.columns:
    df[column] = encoder.fit_transform(df[column])

df

In [ ]:
TRAIN = df.drop('hadheartattack', axis=1)
TARGET = df['hadheartattack']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(TRAIN, TARGET, test_size=0.2, random_state=42, stratify=df['hadheartattack'])
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

In [ ]:
classifier = xgb.XGBClassifier(n_estimators=1000, learning_rate=0.05)
from sklearn.model_selection import cross_val_score

# Perform cross-validation
scores = cross_val_score(classifier, TRAIN, TARGET, cv=3)

# Print the cross-validation scores
print("Cross-Validation Scores:", scores)
print("Mean Cross-Validation Score:", scores.mean())

In [ ]:
# from sklearn.model_selection import GridSearchCV
# import xgboost as xgb

# # Define the parameter grid
# param_grid = {
#     'scale_pos_weight': [1, 10, 25, 50, 75, 99, 100]
# }

# # Create the XGBoost classifier
# classifier = xgb.XGBClassifier(n_estimators=1000, learning_rate=0.05)

# # Create the GridSearchCV object
# grid_search = GridSearchCV(classifier, param_grid, cv=3)

# # Fit the grid search to the training data
# grid_search.fit(X_train, y_train)

# # Print the best parameters and score
# print("Best Parameters:", grid_search.best_params_)
# print("Best Score:", grid_search.best_score_)


In [ ]:
classifier.fit(X_train, y_train, eval_set=[(X_train, y_train), (X_test, y_test)], early_stopping_rounds=50, verbose=True)
classifier.score(X_test, y_test)

In [ ]:
from sklearn.metrics import classification_report

y_pred = classifier.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)


In [ ]:
classifier.feature_importances_

In [ ]:
feature_importances = pd.DataFrame(classifier.feature_importances_, index = X_train.columns, columns=['importance']).sort_values('importance', ascending=False)

In [ ]:
feature_importances.iloc[:10].plot(kind='barh', title='Feature Importance')

# Machine Learning Models

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

models = [
    RandomForestClassifier(random_state=42),
    LogisticRegression(random_state=42),
    DecisionTreeClassifier(random_state=42)
]

for model in models:
    model.fit(X_train, y_train)
    score = model.score(X_test, y_test)
    print(f"{type(model).__name__} Score: {score}")



In [ ]:
from sklearn.ensemble import VotingClassifier

# Instantiate the individual models
voting = VotingClassifier(estimators=[('Random Forest', RandomForestClassifier(random_state=42)),
                                      ('Logistic Regression', LogisticRegression(random_state=42)),
                                      ('Decision Tree', DecisionTreeClassifier(random_state=42))],
                          voting='hard', n_jobs=-1, verbose=True)
voting.fit_transform(X_train, y_train)
voting.score(X_test, y_test)

In [ ]:
y_pred = voting.predict(X_test)
report = classification_report(y_test, y_pred)
print(report)